File: 04Lab/src/main/java/core/services/UserService.java
package core.services;

import core.DTO.request.UserRequest;
import core.DTO.response.UserResponse;
import core.entity.UserEntity;
import core.mapper.UserMapper;
import core.repository.UserRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor

public class UserService {
    private final UserRepository userRepository;
    private final UserMapper userMapper;
    private final PasswordEncoder passwordEncoder;

    public UserResponse getUser(Long id) {
        UserEntity user = userRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("User not found"));
        return userMapper.toResponse(user);
    }

    public List<UserResponse> getAllUsers() {
        List<UserEntity> users = userRepository.findAll();
        return users.stream()
                .map(userMapper::toResponse)
                .collect(Collectors.toList());
    }

    @Transactional
    public UserResponse createUser(UserRequest request) {
        return createUser(request, false); // По умолчанию не админ
    }

    @Transactional
    public UserResponse createUser(UserRequest request, Boolean isAdmin) {
        if (userRepository.existsByName(request.getName())) {throw new RuntimeException("User with this name already exists");}

        UserEntity user = new UserEntity();
        user.setName(request.getName());
        user.setEmail(request.getEmail());

        user.setPassword(passwordEncoder.encode(request.getPassword()).getBytes());

        user.setIsAdmin(isAdmin != null ? isAdmin : false);

        userRepository.save(user);
        return userMapper.toResponse(user);
    }

    @Transactional
    public UserResponse updateUser(Long id, UserRequest request) {
        return updateUser(id, request, null); // isAdmin не меняется
    }

    @Transactional
    public UserResponse updateUser(Long id, UserRequest request, Boolean isAdmin) {
        UserEntity user = userRepository.findById(id).orElseThrow(() -> new RuntimeException("User not found"));

        if (request.getName() != null) {
            user.setName(request.getName());
        }
        if (request.getEmail() != null) {
            user.setEmail(request.getEmail());
        }
        if (request.getPassword() != null) {

            user.setPassword(passwordEncoder.encode(request.getPassword()).getBytes());
        }
        if (isAdmin != null) {
            user.setIsAdmin(isAdmin);
        }

        userRepository.save(user);
        return userMapper.toResponse(user);
    }

    @Transactional
    public void deleteUser(Long id) {
        if (!userRepository.existsById(id)) {
            throw new RuntimeException("User not found");
        }
        userRepository.deleteById(id);
    }

    @Transactional
    public UserResponse registerUser(UserRequest request) {
        return createUser(request, false);
    }


    @Transactional
    public UserResponse changeUserRole(Long id, Boolean isAdmin) {
        UserEntity user = userRepository.findById(id).orElseThrow(() -> new RuntimeException("User not found"));

        user.setIsAdmin(isAdmin);
        userRepository.save(user);

        return userMapper.toResponse(user);
    }
    public UserResponse findByUsername(String username) {
        UserEntity user = userRepository.findByName(username).orElseThrow(() -> new RuntimeException("User not found"));

        return userMapper.toResponse(user);
    }
}

File: 04Lab/src/frontend/src/services/userService.ts
/**
 * Сервис для работы с пользователями (только для администраторов)
 * Соответствует UserController в Spring
 */
import apiClient from '@/api/client'
import type { User, UserRequest } from '@/types/api'

class UserService {
  /**
   * Получить всех пользователей (только для ADMIN)
   * @returns Список пользователей
   */
  async getAllUsers(): Promise<User[]> {
    const response = await apiClient.get<User[]>('/api/users')
    return response.data
  }

  /**
   * Получить пользователя по ID
   * @param id - ID пользователя
   * @returns Данные пользователя
   */
  async getUser(id: number): Promise<User> {
    const response = await apiClient.get<User>(`/api/users/${id}`)
    return response.data
  }

  /**
   * Создать пользователя (только для ADMIN)
   * @param data - данные пользователя
   * @param isAdmin - создать как администратора
   * @returns Созданный пользователь
   */
  async createUser(data: UserRequest, isAdmin: boolean = false): Promise<User> {
    const endpoint = isAdmin ? '/api/users/admin' : '/api/users'
    const response = await apiClient.post<User>(endpoint, data)
    return response.data
  }

  /**
   * Обновить пользователя
   * @param id - ID пользователя
   * @param data - новые данные
   * @returns Обновленный пользователь
   */
  async updateUser(id: number, data: Partial<UserRequest>): Promise<User> {
    const response = await apiClient.put<User>(`/api/users/${id}`, data)
    return response.data
  }

  /**
   * Изменить роль пользователя (только для ADMIN)
   * @param id - ID пользователя
   * @param isAdmin - назначить/снять права администратора
   * @returns Обновленный пользователь
   */
  async changeUserRole(id: number, isAdmin: boolean): Promise<User> {
    const response = await apiClient.put<User>(`/api/users/${id}/role`, { isAdmin })
    return response.data
  }

  /**
   * Удалить пользователя (только для ADMIN)
   * @param id - ID пользователя
   */
  async deleteUser(id: number): Promise<void> {
    await apiClient.delete(`/api/users/${id}`)
  }

  /**
   * Получить текущего пользователя (по токену)
   * @returns Данные текущего пользователя
   */
  async getCurrentUser(): Promise<User | null> {
    try {
      // Получаем свой профиль через /api/users/{id}
      // Для этого нужен ID текущего пользователя
      // Альтернативно: создать endpoint /api/users/me в Spring
      return null
    } catch {
      return null
    }
  }
}

export default new UserService()